{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **DATA Retrieval and Cleanup**  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Data Retrival**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Input error...\n",
      "function usage: bolddata_retrival [-i <Taxonomy>] [-a] [-c <name of country>] [-f <a file with list of countries and named *countries*>]\n",
      "../data/input/bold_data/diptera_13012021/Algeria-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Angola-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Benin-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Botswana-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Burkina_Faso-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Burundi-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Cameroon-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Cape_Verde-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Central_African_Republic-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Chad-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Comoros-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Cote_dIvoire-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Democratic_republic_of_the_Congo-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Djibouti-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Egypt-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Equatorial_Guinea-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Eritrea-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Ethiopia-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Gabon-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Gambia-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Ghana-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Guinea-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Guinea-Bissau-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Kenya-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Lesotho-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Liberia-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Libya-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Madagascar-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Malawi-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Mali-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Mauritania-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Mauritius-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Morocco-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Mozambique-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Namibia-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Niger-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Nigeria-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Republic_of_the_Congo-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Reunion-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Rwanda-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Sao_Tome_and_Principe-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Senegal-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Seychelles-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Sierra_Leone-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Somalia-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/South_Africa-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/South_Sudan-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Sudan-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Swaziland-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Tanzania-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Togo-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Tunisia-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Uganda-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Zambia-13012021.xml\n",
      "../data/input/bold_data/diptera_13012021/Zimbabwe-13012021.xml\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd /home/kibet/bioinformatics/github/co1_metaanalysis/code/\n",
    "source ./process_all_input_files.sh\n",
    "bolddata_retrival #-i Diptera -f countries \n",
    "#Uncomment the command to download from a list of countries in the file countries\n",
    "ls ../data/input/bold_data/diptera_13012021/*-13012021.xml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Data Extraction - XML to TSV**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Input error...\n",
      "Usage: boldxml2tsv file1.*[file2.* file3.* ...]\n",
      "\n",
      "\tgenerating .tsv files from .xml downloads\n",
      "../data/input/bold_data/diptera_13012021/Algeria-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Angola-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Benin-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Botswana-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Burkina_Faso-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Burundi-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Cameroon-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Cape_Verde-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Central_African_Republic-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Comoros-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Cote_dIvoire-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Democratic_republic_of_the_Congo-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Djibouti-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Egypt-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Equatorial_Guinea-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Ethiopia-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Gabon-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Gambia-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Ghana-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Guinea-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Guinea-Bissau-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Kenya-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Lesotho-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Liberia-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Libya-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Madagascar-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Malawi-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Mali-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Mauritania-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Mauritius-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Morocco-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Mozambique-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Namibia-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Niger-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Nigeria-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Republic_of_the_Congo-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Reunion-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Rwanda-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Sao_Tome_and_Principe-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Senegal-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Seychelles-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Sierra_Leone-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/South_Africa-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Sudan-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Swaziland-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Tanzania-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Togo-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Tunisia-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Uganda-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Zambia-13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/Zimbabwe-13012021.tsv\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd /home/kibet/bioinformatics/github/co1_metaanalysis/code/\n",
    "source ./process_all_input_files.sh\n",
    "boldxml2tsv #../data/input/bold_data/diptera_13012021/*_13012021.xml  \n",
    "#Uncomment the command to run execute\n",
    "ls ../data/input/bold_data/diptera_13012021/*-13012021.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Data Sorting**  \n",
    "This step concatenates all the data (tsv files) into one file and sorts them into Sequence-length dependent groupings: \n",
    "1. diptera_13012021_500to700_data.tsv\n",
    "2. diptera_13012021_650to660_data.tsv\n",
    "3. diptera_13012021_all_data.tsv\n",
    "4. diptera_13012021_NonCOI_data.tsv\n",
    "5. diptera_13012021_Over499_data.tsv\n",
    "6. diptera_13012021_Over700_data.tsv\n",
    "7. diptera_13012021_Under500_data.tsv\n",
    "\n",
    "This will reduce complexities in downstream steps like Mustiple Sequence Aligment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Input error...\n",
      "Usage: boldtsv_cleanup file1.*[file2.* file3.* ...]\n",
      "../data/input/bold_data/diptera_13012021/diptera_13012021_500to700_data.fasta\n",
      "../data/input/bold_data/diptera_13012021/diptera_13012021_500to700_data.tsv\n",
      "../data/input/bold_data/diptera_13012021/diptera_13012021_650to660_data.fasta\n",
      "../data/input/bold_data/diptera_13012021/diptera_13012021_650to660_data.tsv\n",
      "../data/input/bold_data/diptera_13012021/diptera_13012021_all_data.fasta\n",
      "../data/input/bold_data/diptera_13012021/diptera_13012021_all_data.genbank_ids\n",
      "../data/input/bold_data/diptera_13012021/diptera_13012021_all_data.tsv\n",
      "../data/input/bold_data/diptera_13012021/diptera_13012021_cleanup\n",
      "../data/input/bold_data/diptera_13012021/diptera_13012021_NonCOI_data.fasta\n",
      "../data/input/bold_data/diptera_13012021/diptera_13012021_NonCOI_data.tsv\n",
      "../data/input/bold_data/diptera_13012021/diptera_13012021_Over499_data.fasta\n",
      "../data/input/bold_data/diptera_13012021/diptera_13012021_Over499_data.tsv\n",
      "../data/input/bold_data/diptera_13012021/diptera_13012021_Over700_data.fasta\n",
      "../data/input/bold_data/diptera_13012021/diptera_13012021_Over700_data.tsv\n",
      "../data/input/bold_data/diptera_13012021/diptera_13012021.tsv\n",
      "../data/input/bold_data/diptera_13012021/diptera_13012021_Under500_data.fasta\n",
      "../data/input/bold_data/diptera_13012021/diptera_13012021_Under500_data.tsv\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd /home/kibet/bioinformatics/github/co1_metaanalysis/code/\n",
    "source ./process_all_input_files.sh\n",
    "#awk ' FNR==NR{print $0}; FNR != NR{ if ( $0 ~ /^processid.*$/) {next} else{print $0}}' \\\n",
    "#../data/input/bold_data/diptera_13012021/*-13012021.tsv >> \\\n",
    "#../data/input/bold_data/diptera_13012021/diptera-13012021.tsv\n",
    "boldtsv_cleanup #diptera-13012021.tsv\n",
    "ls ../data/input/bold_data/diptera_13012021/diptera_13012021*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Data coversion .tsv to .fasta**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Input error...\n",
      "Usage: boldtsv2fasta file1.*[file2.* file3.* ...]\n",
      "generating .fasta files from .tsv metadata files\n",
      "../data/input/bold_data/diptera_13012021/diptera_13012021_500to700_data.fasta\n",
      "../data/input/bold_data/diptera_13012021/diptera_13012021_650to660_data.fasta\n",
      "../data/input/bold_data/diptera_13012021/diptera_13012021_all_data.fasta\n",
      "../data/input/bold_data/diptera_13012021/diptera_13012021_NonCOI_data.fasta\n",
      "../data/input/bold_data/diptera_13012021/diptera_13012021_Over499_data.fasta\n",
      "../data/input/bold_data/diptera_13012021/diptera_13012021_Over700_data.fasta\n",
      "../data/input/bold_data/diptera_13012021/diptera_13012021_Under500_data.fasta\n",
      "../data/input/diptera_13012021/diptera_13012021_all_data.fasta\n",
      "../data/input/diptera_13012021/diptera_13012021_all_data.genbank_ids\n",
      "../data/input/diptera_13012021/diptera_13012021_all_data.tsv\n",
      "../data/input/diptera_13012021/diptera_13012021_all_data_unpublished_Diptera_dgpd.fasta\n",
      "../data/input/diptera_13012021/diptera_13012021_all_data_unpublished_Diptera_dgpd_orig.fasta\n",
      "../data/input/diptera_13012021/diptera_13012021_all_data_unpublished_Diptera_dgpd_subtree.fasta\n",
      "../data/input/diptera_13012021/diptera_13012021_all_data_unpublished_Diptera_dgpd_subtree.hdrs\n",
      "../data/input/diptera_13012021/diptera_13012021_all_data_unpublished_Diptera_dgpd_subtree.ids\n",
      "../data/input/diptera_13012021/diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa\n",
      "../data/input/diptera_13012021/diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_clash_species.tsv\n",
      "../data/input/diptera_13012021/diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_clash.tsv\n",
      "../data/input/diptera_13012021/diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_evaluation\n",
      "../data/input/diptera_13012021/diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_families.tsv\n",
      "../data/input/diptera_13012021/diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_genera.tsv\n",
      "../data/input/diptera_13012021/diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_l-500-Sc-1_headers\n",
      "../data/input/diptera_13012021/diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_species.tsv\n",
      "../data/input/diptera_13012021/diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa.tsv\n",
      "../data/input/diptera_13012021/diptera_13012021_cleanup\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd /home/kibet/bioinformatics/github/co1_metaanalysis/code/\n",
    "source ./process_all_input_files.sh\n",
    "boldtsv2fasta #../data/input/bold_data/diptera_13012021/diptera_13012021*_data.tsv\n",
    "#Uncomment the command to run execute\n",
    "ls ../data/input/bold_data/diptera_13012021/diptera_13012021*.fasta\n",
    "#cp ../data/input/bold_data/diptera_13012021/diptera_13012021_* ../data/input/diptera_13012021/\n",
    "ls ../data/input/diptera_13012021/diptera_13012021_*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **Processing the Unpublished Sequences** \n",
    "### **To retrive unpublished data from [BOLD Systems](http://www.boldsystems.org/index.php/MAS_Management_UserConsole)**, first create a [BOLD systems account](http://www.boldsystems.org/index.php/MAS_Management_NewUserApp), [login](http://www.boldsystems.org/index.php/Login/page?destination=MAS_Management_UserConsole) and request data managers, to share their data sets.  \n",
    "My list of shared data sets are:\n",
    "1. [DS-KENFRUIT](http://www.boldsystems.org/index.php/MAS_Management_DataConsole?codes=DS-KENFRUIT): managed by Dr Scott E. Miller, has 1,427 records  \n",
    "2. [DS-MPALALEP](http://www.boldsystems.org/index.php/MAS_Management_DataConsole?codes=DS-MPALALEP): managed by Dr Scott E. Miller, has 2,472 records  \n",
    "3. [DS-TBILE](http://www.boldsystems.org/index.php/MAS_Management_DataConsole?codes=DS-TBILE) (Now publicly released now): managed by Dr Scott E. Miller has 90 records  \n",
    "\n",
    "My list of container Projects; these contains multiple data sets within them:\n",
    "4. [IDRCK](http://www.boldsystems.org/index.php/MAS_Management_DataConsole?codes=IDRCK): Has a number of subprojects; IDRC,HIVE, KBIR, KALG, KFISH, KPLA, ARAK and KINS. Has 2,110 sequences (COI-5P=1,704, matK=139, rbcLa=267) out of 6,016 specimen and is managed by Dr. Daniel Masiga.  \n",
    "5. [GMTAH](http://www.boldsystems.org/index.php/MAS_Management_DataConsole?codes=GMTAH),[GMTAI](http://www.boldsystems.org/index.php/MAS_Management_DataConsole?codes=GMTAI) and [GMTAJ](http://www.boldsystems.org/index.php/MAS_Management_DataConsole?codes=GMTAJ) projects. All under the Global Malaise Program and the three have a combined total of 60 Projects and 49246 Specimens.  \n",
    ">1. GMTAH: Has 26 projects titled \"Kenya Malaise Mpala 2014\" with 25,514 specimen, 170 species and 21,742 sequences (COI-5P=21,737, 28S=4 and EF1-alpha=1)  \n",
    ">2. GMTAI: Has 26 projects titled \"Kenya Malaise Kinondo 2014\" with 13,656 specimen, 57 species and 11,805 sequences (COI-5P=11,801, 28S=3 and EF1-alpha=1)  \n",
    ">3. GMTAJ: Has 5 projects titled \"Kenya Malaise Turkana 2014\" with 10,076 specimen, 63 species and 5,175 sequences (COI-5P=5,175)  \n",
    "\n",
    "To retrive this data, I logged into the [BOLD Systems MAS management interface](http://www.boldsystems.org/index.php/MAS_Management_UserConsole) through Chromium web browser and for each named project above: DS-KENFRUIT, DS-MPALALEP, DS-TBILE, IDRCK AND GMTAH-GMTAI-GMTAJ (Mpala_Kinondo_Turkana_Malaise_traps), downloaded the spreadsheet and the sequence files to \"/co1_metaanalysis/data/input/input_data/unpublished\" directory.\n",
    "\n",
    "### **Below we proceed with processing the unpublished records**  \n",
    "1. Generate an assamblage of all the above as \"unpublished.fasta\".\n",
    "2. Filter the records from order diptera \"Diptera\" into \"unpublished_Diptera.fasta\".\n",
    "3. Unalign i.e find and replace the \"-\"s from the sequences.\n",
    "4. Put together the diptera_13012021_all_data.fasta and unpublished_Diptera_dgpd.fasta into one file: diptera_13012021_all_data_unpublished_Diptera_dgpd.fasta\n",
    "5. Remove duplicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Input error...\n",
      "Usage: filter_seqs file1.*[file2.* file3.* ...]\n",
      "Input error...\n",
      "Usage: remove_gaps file1.*[file2.* file3.* ...]\n",
      "./unpublished_Diptera_dgpd.fasta\n",
      "./unpublished_Diptera_dgpd_orig.fasta\n",
      "./unpublished_Diptera.fasta\n",
      "./unpublished.fasta\n",
      "Input error...\n",
      "Usage: delete_repeats file1.*[file2.* file3.* ...]\n",
      "alignment\n",
      "classify.nf\n",
      "classify.sbatch\n",
      "coiclassify.sbatch\n",
      "diptera_13012021_all_data.fasta\n",
      "diptera_13012021_all_data.genbank_ids\n",
      "diptera_13012021_all_data.tsv\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd.fasta\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_orig.fasta\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_subtree.fasta\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_subtree.hdrs\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_subtree.ids\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_clash_species.tsv\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_clash.tsv\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_evaluation\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_families.tsv\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_genera.tsv\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_l-500-Sc-1_headers\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_species.tsv\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa.tsv\n",
      "diptera_13012021_cleanup\n",
      "diptera_all.fasta\n",
      "msa.sbatch\n",
      "RDPClassification\n",
      "slurm-714195.out\n",
      "slurm-715184.out\n",
      "slurm-715359.out\n",
      "slurm-715371.out\n",
      "slurm-716105.out\n",
      "subhrds.nf\n",
      "unpublished_Diptera_dgpd.fasta\n",
      "unpublished_Diptera_dgpd_orig.fasta\n",
      "unpublished_Diptera.fasta\n",
      "unpublished.fasta\n",
      "unpublished_ids\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "bash: line 15: return: can only `return' from a function or sourced script\n",
      "bash: line 22: return: can only `return' from a function or sourced script\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd /home/kibet/bioinformatics/github/co1_metaanalysis/data/input/diptera_13012021/\n",
    "source ../../../code/process_all_input_files.sh\n",
    "\n",
    "#1.\n",
    "#cp ../unpublished/unpublished.fasta ./\n",
    "\n",
    "#2.\n",
    "filter_seqs #unpublished.fasta <EOF\n",
    "#3\n",
    "#Diptera\n",
    "#n\n",
    "#EOF\n",
    "\n",
    "#3.\n",
    "remove_gaps #unpublished_Diptera.fasta\n",
    "ls ./unpublished*.fasta\n",
    "\n",
    "#4.\n",
    "#cat ./diptera_13012021_all_data.fasta ./unpublished_Diptera_dgpd.fasta #>> diptera_13012021_all_data_unpublished_Diptera_dgpd.fasta\n",
    "\n",
    "#5.\n",
    "delete_repeats #diptera_13012021_all_data_unpublished_Diptera_dgpd.fasta <<EOF\n",
    "#y\n",
    "#EOF\n",
    "##...11226 records are repeated in diptera_13012021_all_data_unpublished_Diptera_dgpd.fasta\n",
    "ls ./"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Taxonomic Classification**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Input error...\n",
      "Usage: RDPclassifyCOI -i 'file1.fasta* [file2.fasta* file3.fasta* ...]' [-l <INTEGER>] [-m <INTEGERg|G|m|M>]\n",
      "alignment\n",
      "classify.nf\n",
      "classify.sbatch\n",
      "coiclassify.sbatch\n",
      "diptera_13012021_all_data.fasta\n",
      "diptera_13012021_all_data.genbank_ids\n",
      "diptera_13012021_all_data.tsv\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd.fasta\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_orig.fasta\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_subtree.fasta\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_subtree.hdrs\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_subtree.ids\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_clash_species.tsv\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_clash.tsv\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_evaluation\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_families.tsv\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_genera.tsv\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_l-500-Sc-1_headers\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_species.tsv\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa.tsv\n",
      "diptera_13012021_cleanup\n",
      "diptera_all.fasta\n",
      "msa.sbatch\n",
      "RDPClassification\n",
      "slurm-714195.out\n",
      "slurm-715184.out\n",
      "slurm-715359.out\n",
      "slurm-715371.out\n",
      "slurm-716105.out\n",
      "subhrds.nf\n",
      "unpublished_Diptera_dgpd.fasta\n",
      "unpublished_Diptera_dgpd_orig.fasta\n",
      "unpublished_Diptera.fasta\n",
      "unpublished.fasta\n",
      "unpublished_ids\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd /home/kibet/bioinformatics/github/co1_metaanalysis/data/input/diptera_13012021/\n",
    "source ../../../code/classify.sh\n",
    "#Classification using the ILRI hpc\n",
    "#$ sbatch -w compute05 classify.sbatch\n",
    "#Alternatively\n",
    "RDPclassifyCOI #-i diptera_13012021_all_data_unpublished_Diptera_dgpd.fasta -m 4G -l 3000\n",
    "ls "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Evaluation of classification results**  \n",
    "1. Convertion of data\n",
    "2. Evaluation of taxonomic classifictaion results using R\n",
    "3. Generation of FASTA format headers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Input error...\n",
      "Usage: RDPcoiresults2tsv file1.*[file2.* file3.* ...]\n",
      "Input error...\n",
      "Usage: RDPresults_eval file1.tsv|_taxa[file2.* file3.* ...]\n",
      "Error 1 occured on 8\n",
      "Input error...\n",
      "Usage: RDPcoiheaders -i 'file1.tsv|_taxa[file2.* file3.* ...]' [-l <INTEGER>] [-s <percentage decimals>]\n",
      "Error 1 occured on 11\n",
      "alignment\n",
      "classify.nf\n",
      "classify.sbatch\n",
      "coiclassify.sbatch\n",
      "diptera_13012021_all_data.fasta\n",
      "diptera_13012021_all_data.genbank_ids\n",
      "diptera_13012021_all_data.tsv\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd.fasta\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_orig.fasta\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_subtree.fasta\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_subtree.hdrs\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_subtree.ids\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_clash_species.tsv\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_clash.tsv\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_evaluation\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_families.tsv\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_genera.tsv\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_l-500-Sc-1_headers\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_species.tsv\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa.tsv\n",
      "diptera_13012021_cleanup\n",
      "diptera_all.fasta\n",
      "msa.sbatch\n",
      "RDPClassification\n",
      "slurm-714195.out\n",
      "slurm-715184.out\n",
      "slurm-715359.out\n",
      "slurm-715371.out\n",
      "slurm-716105.out\n",
      "subhrds.nf\n",
      "unpublished_Diptera_dgpd.fasta\n",
      "unpublished_Diptera_dgpd_orig.fasta\n",
      "unpublished_Diptera.fasta\n",
      "unpublished.fasta\n",
      "unpublished_ids\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "bash: line 8: return: can only `return' from a function or sourced script\n",
      "bash: line 11: return: can only `return' from a function or sourced script\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd /home/kibet/bioinformatics/github/co1_metaanalysis/data/input/diptera_13012021/\n",
    "source ../../../code/classify.sh\n",
    "\n",
    "#1. Convertion of data\n",
    "RDPcoiresults2tsv #diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa\n",
    "\n",
    "#2. Evaluation of taxonomic classifictaion results using R\n",
    "RDPresults_eval #diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa.tsv\n",
    "\n",
    "#3. Generation of FASTA format headers\n",
    "RDPcoiheaders #-i diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa.tsv -l 500 -s 1\n",
    "\n",
    "#4. Substituting headers in the final FASTA file\n",
    "\n",
    "ls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Processing the Final FASTA file before Multiple Sequence Alignment.\n",
    "1. Substituting headers in the final FASTA file\n",
    "2. Subsetting the file by sequence length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Input error...\n",
      "\tUsage: substitute_hdrs -i 'file1.fasta* [file2.fasta* file3.fasta* ...]' [-h <headers>]\n",
      "Input error...\n",
      "Usage: subset_seqs file1.*[file2.* file3.* ...]\n",
      "alignment\n",
      "classify.nf\n",
      "classify.sbatch\n",
      "coiclassify.sbatch\n",
      "diptera_13012021_all_data.fasta\n",
      "diptera_13012021_all_data.genbank_ids\n",
      "diptera_13012021_all_data.tsv\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd.fasta\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_orig.fasta\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_subtree.fasta\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_subtree.hdrs\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_subtree.ids\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_clash_species.tsv\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_clash.tsv\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_evaluation\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_families.tsv\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_genera.tsv\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_l-500-Sc-1_headers\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_species.tsv\n",
      "diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa.tsv\n",
      "diptera_13012021_cleanup\n",
      "diptera_all.fasta\n",
      "msa.sbatch\n",
      "RDPClassification\n",
      "slurm-714195.out\n",
      "slurm-715184.out\n",
      "slurm-715359.out\n",
      "slurm-715371.out\n",
      "slurm-716105.out\n",
      "subhrds.nf\n",
      "unpublished_Diptera_dgpd.fasta\n",
      "unpublished_Diptera_dgpd_orig.fasta\n",
      "unpublished_Diptera.fasta\n",
      "unpublished.fasta\n",
      "unpublished_ids\n",
      "All_diptera_all_500to700_data-650to660.fasta\n",
      "All_diptera_all.fasta\n",
      "diptera_all_500to700_data-650to660_10000-15000.fasta\n",
      "diptera_all_500to700_data-650to660_15000-20000.fasta\n",
      "diptera_all_500to700_data-650to660_1-5000.fasta\n",
      "diptera_all_500to700_data-650to660_20000-25000.fasta\n",
      "diptera_all_500to700_data-650to660_25000-30000.fasta\n",
      "diptera_all_500to700_data-650to660_30000-35000.fasta\n",
      "diptera_all_500to700_data-650to660_35000-40000.fasta\n",
      "diptera_all_500to700_data-650to660_40000-45000.fasta\n",
      "diptera_all_500to700_data-650to660_45000-50000.fasta\n",
      "diptera_all_500to700_data-650to660_50000-55000.fasta\n",
      "diptera_all_500to700_data-650to660_5000-10000.fasta\n",
      "diptera_all_500to700_data-650to660_55000-60000.fasta\n",
      "diptera_all_500to700_data-650to660_60000-65000.fasta\n",
      "diptera_all_500to700_data-650to660_65000-67954.fasta\n",
      "diptera_all_500to700_data-650to660.fasta\n",
      "diptera_all_650to660.fasta\n",
      "diptera_all.fasta\n",
      "diptera_all_Over700.fasta\n",
      "diptera_all_Under500.fasta\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘alignment’: File exists\n",
      "bash: line 6: return: can only `return' from a function or sourced script\n",
      "mv: missing file operand\n",
      "Try 'mv --help' for more information.\n",
      "bash: line 7: return: can only `return' from a function or sourced script\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd /home/kibet/bioinformatics/github/co1_metaanalysis/data/input/diptera_13012021/\n",
    "source ../../../code/process_all_input_files.sh\n",
    "substitute_hdrs #-i diptera_all.fasta -h diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa_l-500-Sc-1_headers\n",
    "#Sub_setting by Seq. length\n",
    "subset_seqs #diptera_all.fasta\n",
    "mkdir alignment \n",
    "mv #diptera_all_* ./alignment/ && cp diptera_all.fasta ./alignment/\n",
    "ls ./\n",
    "ls ./alignment/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pre-MSA Processing, Multiple Sequence Alignment (MSA) and Post_MSA Processing.\n",
    "1. Subsetting large FASTA files: diptera_all_500to700_data-650to660.fasta (67954 Sequences)\n",
    "2. Alignment using PASTA\n",
    "3. Visualise the alignments in seaview, identify poorly aligned files, subset the poorly aligned files\n",
    "4. Refining all alignments using MUSCLE (`- refine` option)\n",
    "5. Visualise the alignments in seaview and edit the alignments manually where necessary\n",
    "6. Merge all alignment files using MUSCLE (`-profile` option)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "diptera_all_500to700_data-650to660_10000-15000.aln\n",
      "diptera_all_500to700_data-650to660_10000-15000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_10000-15000_rfnd_diptera_all_500to700_data-650to660_15000-20000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_10000-20000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_1-10000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_15000-20000.aln\n",
      "diptera_all_500to700_data-650to660_15000-20000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_1-5000.aln\n",
      "diptera_all_500to700_data-650to660_1-5000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_1-5000_rfnd_diptera_all_500to700_data-650to660_5000-10000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_20000-25000.aln\n",
      "diptera_all_500to700_data-650to660_20000-25000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_20000-25000_rfnd_diptera_all_500to700_data-650to660_25000-30000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_20000-30000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_25000-30000.aln\n",
      "diptera_all_500to700_data-650to660_25000-30000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_30000-35000.aln\n",
      "diptera_all_500to700_data-650to660_30000-35000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_30000-35000_rfnd_diptera_all_500to700_data-650to660_35000-40000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_30000-40000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_35000-40000.aln\n",
      "diptera_all_500to700_data-650to660_35000-40000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_40000-45000.aln\n",
      "diptera_all_500to700_data-650to660_40000-45000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_40000-45000_rfnd_diptera_all_500to700_data-650to660_45000-50000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_40000-50000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_45000-50000.aln\n",
      "diptera_all_500to700_data-650to660_45000-50000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_50000-51000.aln\n",
      "diptera_all_500to700_data-650to660_50000-51000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_50000-51000_rfnd_diptera_all_500to700_data-650to660_51000-52000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_50000-52000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_50000-52000_rfnd_diptera_all_500to700_data-650to660_52000-54000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_50000-54000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_50000-54000_rfnd_diptera_all_500to700_data-650to660_54000-55000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_50000-55000.aln\n",
      "diptera_all_500to700_data-650to660_50000-55000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_50000-55000_rfnd_rfnd_diptera_all_500to700_data-650to660_55000-60000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_50000-60000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_5000-10000.aln\n",
      "diptera_all_500to700_data-650to660_5000-10000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_51000-52000.aln\n",
      "diptera_all_500to700_data-650to660_51000-52000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_52000-53000.aln\n",
      "diptera_all_500to700_data-650to660_52000-53000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_52000-53000_rfnd_diptera_all_500to700_data-650to660_53000-54000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_52000-54000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_53000-54000.aln\n",
      "diptera_all_500to700_data-650to660_53000-54000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_54000-55000.aln\n",
      "diptera_all_500to700_data-650to660_54000-55000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_55000-60000.aln\n",
      "diptera_all_500to700_data-650to660_55000-60000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_60000-65000.aln\n",
      "diptera_all_500to700_data-650to660_60000-65000_rfnd.afa\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "bash: line 2: ../../../code/align.sh: No such file or directory\n",
      "mv: cannot stat 'diptera_all_500to700_data-650to660.fasta': No such file or directory\n",
      "bash: line 67: muscle_refine: command not found\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd /home/kibet/bioinformatics/github/co1_metaanalysis/data/input/diptera_13012021/alignment/\n",
    "source ../../../code/align.sh\n",
    "\n",
    "##1. Subsetting diptera_all_500to700_data-650to660.fasta\n",
    "#head -n 5000 diptera_all_650to660.fasta > diptera_all_500to700_data-650to660_1-5000.fasta\n",
    "#head -n 10000 diptera_all_650to660.fasta | tail -n 5000 diptera_all_650to660.fasta > diptera_all_500to700_data-650to660_5000-10000.fasta\n",
    "#head -n 15000 diptera_all_650to660.fasta | tail -n 5000 diptera_all_650to660.fasta > diptera_all_500to700_data-650to660_10000-15000.fasta\n",
    "#head -n 20000 diptera_all_650to660.fasta | tail -n 5000 diptera_all_650to660.fasta > diptera_all_500to700_data-650to660_15000-20000.fasta\n",
    "#head -n 25000 diptera_all_650to660.fasta | tail -n 5000 diptera_all_650to660.fasta > diptera_all_500to700_data-650to660_20000-25000.fasta\n",
    "#head -n 30000 diptera_all_650to660.fasta | tail -n 5000 diptera_all_650to660.fasta > diptera_all_500to700_data-650to660_25000-30000.fasta\n",
    "#head -n 35000 diptera_all_650to660.fasta | tail -n 5000 diptera_all_650to660.fasta > diptera_all_500to700_data-650to660_30000-35000.fasta\n",
    "#head -n 40000 diptera_all_650to660.fasta | tail -n 5000 diptera_all_650to660.fasta > diptera_all_500to700_data-650to660_35000-40000.fasta\n",
    "#head -n 45000 diptera_all_650to660.fasta | tail -n 5000 diptera_all_650to660.fasta > diptera_all_500to700_data-650to660_40000-45000.fasta\n",
    "#head -n 50000 diptera_all_650to660.fasta | tail -n 5000 diptera_all_650to660.fasta > diptera_all_500to700_data-650to660_45000-50000.fasta\n",
    "#head -n 55000 diptera_all_650to660.fasta | tail -n 5000 diptera_all_650to660.fasta > diptera_all_500to700_data-650to660_50000-55000.fasta\n",
    "#head -n 60000 diptera_all_650to660.fasta | tail -n 5000 diptera_all_650to660.fasta > diptera_all_500to700_data-650to660_55000-60000.fasta\n",
    "#head -n 65000 diptera_all_650to660.fasta | tail -n 5000 diptera_all_650to660.fasta > diptera_all_500to700_data-650to660_60000-65000.fasta\n",
    "#tail -n 2954 diptera_all_650to660.fasta > diptera_all_500to700_data-650to660_65000-67954.fasta\n",
    "mv diptera_all_500to700_data-650to660.fasta All_diptera_all_500to700_data-650to660.fasta\n",
    "\n",
    "##2. Alignment using PASTA\n",
    "#pasta_aln /home/gkibet/bioinformatics/github/co1_metaanalysis/data/input/diptera_13012021/alignment/diptera_all_*.fasta <<EOF\n",
    "#1\n",
    "#16\n",
    "#1\n",
    "#16\n",
    "#1\n",
    "#16\n",
    "#1\n",
    "#16\n",
    "#1\n",
    "#16\n",
    "#1\n",
    "#16\n",
    "#1\n",
    "#16\n",
    "#1\n",
    "#16\n",
    "#1\n",
    "#16\n",
    "#1\n",
    "#16\n",
    "#1\n",
    "#16\n",
    "#1\n",
    "#16\n",
    "#1\n",
    "#16\n",
    "#1\n",
    "#16\n",
    "#1\n",
    "#16\n",
    "#1\n",
    "#16\n",
    "#1\n",
    "#16\n",
    "#EOF\n",
    "\n",
    "cd /home/kibet/bioinformatics/github/co1_metaanalysis/data/output/alignment/pasta_output/aligned/diptera_13012021/\n",
    "##3. Of all MSA outputs diptera_all_500to700_data-650to660_50000-55000.aln had a poor alignment.\n",
    "#it was devided into 1000 sequence files and refined using MUSCLE:\n",
    "#head -n 2000 diptera_all_500to700_data-650to660_50000-55000.aln > diptera_all_500to700_data-650to660_50000-51000.aln\n",
    "#head -n 4000 diptera_all_500to700_data-650to660_50000-55000.aln | tail -n 2000 > diptera_all_500to700_data-650to660_51000-52000.aln\n",
    "#head -n 6000 diptera_all_500to700_data-650to660_50000-55000.aln | tail -n 2000 > diptera_all_500to700_data-650to660_52000-53000.aln\n",
    "#head -n 8000 diptera_all_500to700_data-650to660_50000-55000.aln | tail -n 2000 > diptera_all_500to700_data-650to660_53000-54000.aln\n",
    "#head -n 10000 diptera_all_500to700_data-650to660_50000-55000.aln | tail -n 2000 > diptera_all_500to700_data-650to660_54000-55000.aln\n",
    "muscle_refine #diptera_all_*000.aln\n",
    "ls diptera_all_*000.aln diptera_all_*000_rfnd.afa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "6. Merge all alignment files using MUSCLE (`-profile` option)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "diptera_all_500to700_data-650to660_10000-15000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_15000-20000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_1-5000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_20000-25000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_25000-30000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_30000-35000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_35000-40000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_40000-45000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_45000-50000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_50000-51000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_50000-51000_rfnd_diptera_all_500to700_data-650to660_51000-52000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_50000-52000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_50000-52000_rfnd_diptera_all_500to700_data-650to660_52000-54000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_50000-54000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_50000-54000_rfnd_diptera_all_500to700_data-650to660_54000-55000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_50000-55000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_50000-55000_rfnd_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_5000-10000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_51000-52000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_52000-53000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_52000-53000_rfnd_diptera_all_500to700_data-650to660_53000-54000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_52000-54000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_53000-54000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_54000-55000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_55000-60000_rfnd.afa\n",
      "diptera_all_500to700_data-650to660_60000-65000_rfnd.afa\n",
      "diptera_all_650to660_rfnd.afa\n",
      "diptera_all_Over700_rfnd.afa\n",
      "diptera_all_Under500_rfnd.afa\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "bash: line 2: ../../../code/align.sh: No such file or directory\n",
      "bash: line 5: muscle_p2p: command not found\n",
      "bash: line 7: muscle_p2p: command not found\n",
      "bash: line 9: muscle_p2p: command not found\n",
      "bash: line 11: muscle_p2p: command not found\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd /home/kibet/bioinformatics/github/co1_metaanalysis/data/output/alignment/pasta_output/aligned/diptera_13012021/\n",
    "source ../../../code/align.sh\n",
    "##Edited the alignments in Seaview.\n",
    "##Merge all alignment files\n",
    "muscle_p2p #diptera_all_500to700_data-650to660_50000-51000_rfnd.afa diptera_all_500to700_data-650to660_51000-52000_rfnd.afa\n",
    "cp diptera_all_500to700_data-650to660_50000-51000_rfnd_diptera_all_500to700_data-650to660_51000-52000_rfnd.afa diptera_all_500to700_data-650to660_50000-52000_rfnd.afa\n",
    "muscle_p2p #diptera_all_500to700_data-650to660_52000-53000_rfnd.afa diptera_all_500to700_data-650to660_53000-54000_rfnd.afa\n",
    "cp diptera_all_500to700_data-650to660_52000-53000_rfnd_diptera_all_500to700_data-650to660_53000-54000_rfnd.afa diptera_all_500to700_data-650to660_52000-54000_rfnd.afa\n",
    "muscle_p2p #diptera_all_500to700_data-650to660_50000-52000_rfnd.afa diptera_all_500to700_data-650to660_52000-54000_rfnd.afa\n",
    "cp diptera_all_500to700_data-650to660_50000-52000_rfnd_diptera_all_500to700_data-650to660_52000-54000_rfnd.afa diptera_all_500to700_data-650to660_50000-54000_rfnd.afa\n",
    "muscle_p2p #diptera_all_500to700_data-650to660_50000-54000_rfnd.afa diptera_all_500to700_data-650to660_54000-55000_rfnd.afa\n",
    "cp diptera_all_500to700_data-650to660_50000-54000_rfnd_diptera_all_500to700_data-650to660_54000-55000_rfnd.afa diptera_all_500to700_data-650to660_50000-55000_rfnd_rfnd.afa\n",
    "muscle_p2p #diptera_all_500to700_data-650to660_1-5000_rfnd.afa diptera_all_500to700_data-650to660_5000-10000_rfnd.afa\n",
    "cp diptera_all_500to700_data-650to660_1-5000_rfnd_diptera_all_500to700_data-650to660_5000-10000_rfnd.afa diptera_all_500to700_data-650to660_1-10000_rfnd.afa\n",
    "muscle_p2p #diptera_all_500to700_data-650to660_10000-15000_rfnd.afa diptera_all_500to700_data-650to660_15000-20000_rfnd.afa\n",
    "cp diptera_all_500to700_data-650to660_10000-15000_rfnd_diptera_all_500to700_data-650to660_15000-20000_rfnd.afa diptera_all_500to700_data-650to660_10000-20000_rfnd.afa\n",
    "muscle_p2p #diptera_all_500to700_data-650to660_20000-25000_rfnd.afa diptera_all_500to700_data-650to660_25000-30000_rfnd.afa\n",
    "cp diptera_all_500to700_data-650to660_20000-25000_rfnd_diptera_all_500to700_data-650to660_25000-30000_rfnd.afa diptera_all_500to700_data-650to660_20000-30000_rfnd.afa\n",
    "muscle_p2p #diptera_all_500to700_data-650to660_30000-35000_rfnd.afa diptera_all_500to700_data-650to660_35000-40000_rfnd.afa\n",
    "cp diptera_all_500to700_data-650to660_30000-35000_rfnd_diptera_all_500to700_data-650to660_35000-40000_rfnd.afa diptera_all_500to700_data-650to660_30000-40000_rfnd.afa\n",
    "muscle_p2p #diptera_all_500to700_data-650to660_40000-45000_rfnd.afa diptera_all_500to700_data-650to660_45000-50000_rfnd.afa\n",
    "cp diptera_all_500to700_data-650to660_40000-45000_rfnd_diptera_all_500to700_data-650to660_45000-50000_rfnd.afa diptera_all_500to700_data-650to660_40000-50000_rfnd.afa\n",
    "muscle_p2p #diptera_all_500to700_data-650to660_50000-55000_rfnd_rfnd.afa diptera_all_500to700_data-650to660_55000-60000_rfnd.afa\n",
    "cp diptera_all_500to700_data-650to660_50000-55000_rfnd_rfnd_diptera_all_500to700_data-650to660_55000-60000_rfnd.afa diptera_all_500to700_data-650to660_50000-60000_rfnd.afa\n",
    "muscle_p2p #diptera_all_500to700_data-650to660_60000-65000_rfnd.afa diptera_all_500to700_data-650to660_65000-67954.aln\n",
    "cp diptera_all_500to700_data-650to660_60000-65000_rfnd_diptera_all_500to700_data-650to660_65000-67954.afa diptera_all_500to700_data-650to660_60000-67954_rfnd.afa\n",
    "muscle_p2p #diptera_all_500to700_data-650to660_1-10000_rfnd.afa diptera_all_500to700_data-650to660_10000-20000_rfnd.afa\n",
    "cp diptera_all_500to700_data-650to660_1-10000_rfnd_diptera_all_500to700_data-650to660_10000-20000_rfnd.afa diptera_all_500to700_data-650to660_1-20000_rfnd.afa\n",
    "muscle_p2p #diptera_all_500to700_data-650to660_20000-30000_rfnd.afa diptera_all_500to700_data-650to660_30000-40000_rfnd.afa\n",
    "cp diptera_all_500to700_data-650to660_20000-30000_rfnd_diptera_all_500to700_data-650to660_30000-40000_rfnd.afa diptera_all_500to700_data-650to660_20000-40000_rfnd.afa\n",
    "muscle_p2p #diptera_all_500to700_data-650to660_40000-50000_rfnd.afa diptera_all_500to700_data-650to660_50000-60000_rfnd.afa\n",
    "cp diptera_all_500to700_data-650to660_40000-50000_rfnd_diptera_all_500to700_data-650to660_50000-60000_rfnd.afa  diptera_all_500to700_data-650to660_40000-60000_rfnd.afa\n",
    "muscle_p2p #diptera_all_500to700_data-650to660_1-20000_rfnd.afa diptera_all_500to700_data-650to660_20000-40000_rfnd.afa\n",
    "cp diptera_all_500to700_data-650to660_1-20000_rfnd_diptera_all_500to700_data-650to660_20000-40000_rfnd.afa diptera_all_500to700_data-650to660_1-40000_rfnd.afa\n",
    "muscle_p2p #diptera_all_500to700_data-650to660_40000-60000_rfnd.afa diptera_all_500to700_data-650to660_60000-67954_rfnd.afa\n",
    "cp diptera_all_500to700_data-650to660_40000-60000_rfnd_diptera_all_500to700_data-650to660_60000-67954_rfnd.afa diptera_all_500to700_data-650to660_40000-67954_rfnd.afa \n",
    "\n",
    "ls *_rfnd.afa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
