{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **Using the RDP Classifier to add taxonomic classification to the sequences.**  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Taxonomic classification of Diptera sequences using RDP classifier**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./diptera_13012021_all_data_unpublished_Diptera_dgpd_01.fasta\n",
      "./diptera_13012021_all_data_unpublished_Diptera_dgpd_01_taxa\n",
      "./diptera_13012021_all_data_unpublished_Diptera_dgpd_02.fasta\n",
      "./diptera_13012021_all_data_unpublished_Diptera_dgpd_02_taxa\n",
      "./diptera_13012021_all_data_unpublished_Diptera_dgpd_03.fasta\n",
      "./diptera_13012021_all_data_unpublished_Diptera_dgpd_03_taxa\n",
      "./diptera_13012021_all_data_unpublished_Diptera_dgpd_04.fasta\n",
      "./diptera_13012021_all_data_unpublished_Diptera_dgpd_04_taxa\n",
      "./diptera_13012021_all_data_unpublished_Diptera_dgpd_05.fasta\n",
      "./diptera_13012021_all_data_unpublished_Diptera_dgpd_05_taxa\n",
      "./diptera_13012021_all_data_unpublished_Diptera_dgpd_06.fasta\n",
      "./diptera_13012021_all_data_unpublished_Diptera_dgpd_06_taxa\n",
      "./diptera_13012021_all_data_unpublished_Diptera_dgpd_07.fasta\n",
      "./diptera_13012021_all_data_unpublished_Diptera_dgpd_07_taxa\n",
      "./diptera_13012021_all_data_unpublished_Diptera_dgpd_08.fasta\n",
      "./diptera_13012021_all_data_unpublished_Diptera_dgpd_08_taxa\n",
      "./diptera_13012021_all_data_unpublished_Diptera_dgpd_09.fasta\n",
      "./diptera_13012021_all_data_unpublished_Diptera_dgpd_09_taxa\n",
      "./diptera_13012021_all_data_unpublished_Diptera_dgpd_taxa\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "bash: line 6: ../../../code/classify.sh: No such file or directory\n",
      "bash: line 7: RDPclassifyCOI: command not found\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ~/bioinformatics/github/co1_metaanalysis/data/input/diptera_13012021/test01/\n",
    "\n",
    "# Executing RDP classifier, this includes the setting of the maximum size or memeory -Xmx**** to 80g\n",
    "#java -Xmx80g -jar ~/bioinformatics/github/co1_metaanalysis/code/tools/RDPTools/classifier.jar classify -c 0.8 -t ~/bioinformatics/github/co1_metaanalysis/code/tools/CO1Classifierv3.2/mydata_trained/rRNAClassifier.properties -o enafroCOI_all_clean_Diptera_taxa1 enafroCOI_all_clean_Diptera.fasta\n",
    "\n",
    "source ../../../code/classify.sh\n",
    "RDPclassifyCOI #-i ./*.fasta -l 3000 -m 20G\n",
    "ls ./*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Reformating the taxa file to tsv**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Process_ID\tPhylum\tClass\tOrder\tFamily\tSubfamily\tTribe\tGenus\tSpecies\tSubspecies\tCountry\tExactsite\tLatitude\tLongitude\tElevation\tseq_len\tAs_CellularOrganism\tCellularOrganism_sc\tAs_superkingdom\tSuperkingdom_sc\tAs_kingdom\tKingdom_sc\tAs_phylum\tPhylum_sc\tAs_class\tClass_sc\tAs_order\tOrder_sc\tAs_family\tFamily_sc\tAs_genus\tGenus_sc\tAs_species\tSpecies_sc\n",
      "GBDP19110-15\tArthropoda\tInsecta\tDiptera\tPsychodidae\tPhlebotominae\tNA\tSergentomyia\tSergentomyia_minuta\tNA\tAlgeria\tNA\tNA\tNA\tNA\t564\tcellularOrganisms\t1.0\tEukaryota\t1.0\tMetazoa\t1.0\tArthropoda\t1.0\tInsecta\t1.0\tDiptera\t1.0\tPsychodidae\t1.0\tSergentomyia\t1.0\tSergentomyia_minuta\t1.0\n",
      "GBDP19109-15\tArthropoda\tInsecta\tDiptera\tPsychodidae\tPhlebotominae\tNA\tSergentomyia\tSergentomyia_minuta\tNA\tAlgeria\tNA\tNA\tNA\tNA\t564\tcellularOrganisms\t1.0\tEukaryota\t1.0\tMetazoa\t1.0\tArthropoda\t1.0\tInsecta\t1.0\tDiptera\t1.0\tPsychodidae\t1.0\tSergentomyia\t1.0\tSergentomyia_minuta\t1.0\n",
      "GBDP19118-15\tArthropoda\tInsecta\tDiptera\tPsychodidae\tPhlebotominae\tNA\tSergentomyia\tSergentomyia_minuta\tNA\tAlgeria\tNA\tNA\tNA\tNA\t564\tcellularOrganisms\t1.0\tEukaryota\t1.0\tMetazoa\t1.0\tArthropoda\t1.0\tInsecta\t1.0\tDiptera\t1.0\tPsychodidae\t1.0\tSergentomyia\t1.0\tSergentomyia_minuta\t1.0\n",
      "GBDP4645-08\tArthropoda\tInsecta\tDiptera\tCeratopogonidae\tCeratopogoninae\tNA\tCulicoides\tCulicoides_imicola\tNA\tAlgeria\tNA\tNA\tNA\tNA\t471\tcellularOrganisms\t1.0\tEukaryota\t1.0\tMetazoa\t1.0\tArthropoda\t1.0\tInsecta\t1.0\tDiptera\t1.0\tCeratopogonidae\t1.0\tCulicoides\t1.0\tCulicoides_imicola\t1.0\n",
      "GBDP4646-08\tArthropoda\tInsecta\tDiptera\tCeratopogonidae\tCeratopogoninae\tNA\tCulicoides\tCulicoides_imicola\tNA\tAlgeria\tNA\tNA\tNA\tNA\t471\tcellularOrganisms\t1.0\tEukaryota\t1.0\tMetazoa\t1.0\tArthropoda\t1.0\tInsecta\t1.0\tDiptera\t1.0\tCeratopogonidae\t1.0\tCulicoides\t1.0\tCulicoides_imicola\t1.0\n",
      "GBDP4647-08\tArthropoda\tInsecta\tDiptera\tCeratopogonidae\tCeratopogoninae\tNA\tCulicoides\tCulicoides_imicola\tNA\tAlgeria\tNA\tNA\tNA\tNA\t471\tcellularOrganisms\t1.0\tEukaryota\t1.0\tMetazoa\t1.0\tArthropoda\t1.0\tInsecta\t1.0\tDiptera\t1.0\tCeratopogonidae\t1.0\tCulicoides\t1.0\tCulicoides_imicola\t1.0\n",
      "GBMIN46771-17\tArthropoda\tInsecta\tDiptera\tCulicidae\tCulicinae\tNA\tCulex\tCulex_pipiens\tCulex_pipiens_pipiens\tAlgeria\tNA\tNA\tNA\tNA\t450\tcellularOrganisms\t1.0\tEukaryota\t1.0\tMetazoa\t1.0\tArthropoda\t1.0\tInsecta\t1.0\tDiptera\t1.0\tCulicidae\t1.0\tCulex\t1.0\tCulex_terzii\t0.95\n",
      "GBMIN54028-17\tArthropoda\tInsecta\tDiptera\tCeratopogonidae\tCeratopogoninae\tNA\tCulicoides\tCulicoides_cataneii\tNA\tAlgeria\tNA\tNA\tNA\tNA\t472\tcellularOrganisms\t1.0\tEukaryota\t1.0\tMetazoa\t1.0\tArthropoda\t1.0\tInsecta\t1.0\tDiptera\t1.0\tCeratopogonidae\t1.0\tCulicoides\t1.0\tCulicoides_cataneii\t1.0\n",
      "GBMIN54029-17\tArthropoda\tInsecta\tDiptera\tCeratopogonidae\tCeratopogoninae\tNA\tCulicoides\tCulicoides_cataneii\tNA\tAlgeria\tNA\tNA\tNA\tNA\t472\tcellularOrganisms\t1.0\tEukaryota\t1.0\tMetazoa\t1.0\tArthropoda\t1.0\tInsecta\t1.0\tDiptera\t1.0\tCeratopogonidae\t1.0\tCulicoides\t1.0\tCulicoides_cataneii\t1.0\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ~/bioinformatics/github/co1_metaanalysis/data/input/diptera_13012021/test01/\n",
    "source ../../../code/classify.sh\n",
    "\n",
    "RDPcoiresults2tsv enafroCOI_all_clean_Diptera_taxa enafroCOI_diptera_all_data_taxa\n",
    "\n",
    "head -10 enafroCOI_all_clean_Diptera_taxa.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Simple Statistical analysis using R**  \n",
    "This exploits an Rscript  whose details have been explaiined in [RDPclassifier_results.ipynb](./RDPclassifier_results.ipynb)  \n",
    "It Generates 3 files:  \n",
    "1. .*families.tsv - contains a list of all families predifined and assigned at a bootstrap score of 1 and from sequences of >= 500 seqeunce length\n",
    "2. .*genera.tsv - contains a list of all genera predifined and assigned at a bootstrap score of 1 and from sequences of >= 500 seqeunce length\n",
    "3. .*clash.tsv - contains a data frame  of all sequnces with unmatched predefined and assigned species at a bootstrap of 1.0 or sequence length of >= 500\n",
    "\n",
    "\">=\" - greater or equals to"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Checking if needed packages are installed... tidyverse packages: 'ggplot2','dplyr' and 'magrittr'\n",
      "Excellent tidyverse already installed.\n",
      "proceeding with summary analysis of RDPclassifier results...\n",
      "17 predefined families (seq_len >= 500), with frequency of over 20, are found in this clade, out of 54 represented families (seq_len >= 500)\n",
      "\n",
      "16 assigned families (BS score = 1.0 & seq_len >= 500), with frequency of over 20, are found in this clade out of 33 represented assigned families (seq_len >= 500)\n",
      "\n",
      "            Family prefamfreq asfamfreq\n",
      "2         Asilidae         42        40\n",
      "3    Calliphoridae          7        23\n",
      "4    Cecidomyiidae         75         2\n",
      "5  Ceratopogonidae        384       231\n",
      "6     Chironomidae        137        67\n",
      "7        Culicidae        290       413\n",
      "8    Drosophilidae         30        43\n",
      "10       Hybotidae         76        43\n",
      "12        Muscidae         11        60\n",
      "14        Phoridae         28        12\n",
      "16     Psychodidae        152       168\n",
      "18   Sarcophagidae         51        88\n",
      "19       Sciaridae         21         1\n",
      "21      Simuliidae         67        50\n",
      "24       Syrphidae         61       263\n",
      "25       Tabanidae        100       193\n",
      "26      Tachinidae         22        22\n",
      "27     Tephritidae       1092      1100\n",
      "23 predefined genera (seq_len >= 500 $ frequency >= 10) are found in this clade, out of 153 represented genera (seq_len >= 500)\n",
      "\n",
      "37 assigned genera (BS score = 1.0 & seq_len >= 500), with frequency of over 20, are found in this clade out of 142 represented genera (seq_len >= 500)\n",
      "\n",
      "          Genus pregenusfreq asgenusfreq\n",
      "3         Aedes           74          35\n",
      "6     Anopheles           94         177\n",
      "9    Bactrocera          319         291\n",
      "11  Capparimyia           10          22\n",
      "14    Ceratitis          210         288\n",
      "24   Culicoides          317         231\n",
      "25        Dacus          321         345\n",
      "27  Elaphropeza           44          28\n",
      "51      Paragus            1          41\n",
      "54  Phlebotomus           87         109\n",
      "61      Rhingia            2          23\n",
      "62   Sarcophaga           36          53\n",
      "64 Sergentomyia           52          51\n",
      "65     Simulium           52          50\n",
      "71      Tabanus           50         136\n",
      "72   Tanytarsus           16          27\n",
      "74  Trirhithrum           20          49\n",
      "77    Zaprionus           29          29\n",
      "9978 total number of records with 9978 records having over 500 base-pairs\n",
      "2023 records with a predefined species from which 2023 have over 500 base pairs\n",
      "\n",
      "9978 records were assigned species classification with bootstrap support ranging from 0 to 1\n",
      "\n",
      "2246 records were classiffied with a bootsrap of = 1, 2626 with >= 0.95 bootstrap support and 3315 with >= 0.7 bootstrap support\n",
      "\n",
      "1373 records, with predefined species, were classiffied with a bootsrap of = 1, 1635 with >= 0.95 bootstrap support and 1741 with >= 0.7 bootstrap support\n",
      "\n",
      "2246 records, with >= 500 seq. length, were classiffied with a bootsrap of = 1, 2626 with >= 0.95 bootstrap support and 3315 with >= 0.7 bootstrap support\n",
      "\n",
      "1373 records, with predefined species and >= 500 seq. length, were classiffied with a bootsrap of = 1, 1635 with >= 0.95 bootstrap support and 1741 with >= 0.7 bootstrap support\n",
      "\n",
      "2896 records either had predefined species, or were classified with a bootstrap value of 1.0, 3014 with bootstrap value of >=0.95 and 3597 with boostrap value of >=0.7\n",
      "\n",
      "2246 records classified into species have sequences with over 500 bases, with 873 having no prior species classification\n",
      "115 records with a species classication confidence of 1.0 have differing assinged species from predefined species, 115 of which have over 500 base-pairs\n",
      "\n",
      "\n",
      "Checking if needed packages are installed... tidyverse packages: 'ggplot2','dplyr' and 'magrittr'\n",
      "Excellent tidyverse already installed.\n",
      "proceeding with summary analysis of RDPclassifier results...\n",
      "33 predefined families (seq_len >= 500), with frequency of over 20, are found in this clade, out of 62 represented families (seq_len >= 500)\n",
      "\n",
      "22 assigned families (BS score = 1.0 & seq_len >= 500), with frequency of over 20, are found in this clade out of 45 represented assigned families (seq_len >= 500)\n",
      "\n",
      "            Family prefamfreq asfamfreq\n",
      "1      Agromyzidae        232        41\n",
      "2     Anthomyiidae         38        14\n",
      "3         Asilidae         54        52\n",
      "5  Blephariceridae         25        25\n",
      "6    Calliphoridae        459       492\n",
      "7    Cecidomyiidae        345         9\n",
      "8  Ceratopogonidae       2767      1246\n",
      "9     Chironomidae        289       341\n",
      "10     Chloropidae         66         1\n",
      "11       Culicidae       1183      2061\n",
      "12   Drosophilidae         92       116\n",
      "13      Ephydridae         85         6\n",
      "17       Hybotidae        190        77\n",
      "19     Milichiidae         43         4\n",
      "20        Muscidae        686       664\n",
      "23        Phoridae        593       685\n",
      "24    Pipunculidae         47        67\n",
      "26     Psychodidae       1692       636\n",
      "28   Sarcophagidae        284       319\n",
      "30       Sciaridae        119        52\n",
      "31        Sepsidae         95        99\n",
      "32      Simuliidae         80        64\n",
      "33  Sphaeroceridae        224        16\n",
      "35       Syrphidae        129       346\n",
      "36       Tabanidae        127       240\n",
      "37      Tachinidae        113        66\n",
      "38     Tephritidae       1902      1862\n",
      "60 predefined genera (seq_len >= 500 $ frequency >= 10) are found in this clade, out of 249 represented genera (seq_len >= 500)\n",
      "\n",
      "77 assigned genera (BS score = 1.0 & seq_len >= 500), with frequency of over 20, are found in this clade out of 199 represented genera (seq_len >= 500)\n",
      "\n",
      "               Genus pregenusfreq asgenusfreq\n",
      "4              Aedes          546         282\n",
      "6         Allobaccha            1          22\n",
      "9          Anopheles          313        1159\n",
      "13        Atherigona          382          11\n",
      "14          Atylotus           12          23\n",
      "15    Australosepsis           70          73\n",
      "16        Bactrocera          473         435\n",
      "22       Capparimyia           18          30\n",
      "23  Carpophthoromyia           26          18\n",
      "25         Ceratitis          591         659\n",
      "27        Chironomus           52          83\n",
      "28         Chrysomya           61          72\n",
      "32          Coenosia           51         116\n",
      "35    Coquillettidia           19          35\n",
      "37       Corynoptera           20          32\n",
      "38        Cricotopus           43          43\n",
      "39      Crossopalpus           38          36\n",
      "41             Culex           83         378\n",
      "42        Culicoides         1249         962\n",
      "43             Dacus          501         523\n",
      "44         Dasyhelea          121         273\n",
      "48        Drosophila           14          35\n",
      "49       Elaphropeza           46          30\n",
      "50           Elporia           25          25\n",
      "58         Eudorylas           19          26\n",
      "65       Forcipomyia          167           8\n",
      "69       Haematopota            9          22\n",
      "79         Liriomyza           36          22\n",
      "80             Lispe          195           1\n",
      "83          Mansonia           50          96\n",
      "84         Megaselia          280         260\n",
      "92             Musca           31         497\n",
      "102          Paragus            5          58\n",
      "105      Phlebotomus           91         112\n",
      "111         Pollenia          363         373\n",
      "116         Psychoda           41          22\n",
      "119          Rhingia            3          24\n",
      "121       Sarcophaga          134         158\n",
      "125           Sepsis           22          24\n",
      "126     Sergentomyia           84          84\n",
      "127         Simulium           65          64\n",
      "130    Sphaerophoria           25          32\n",
      "138          Tabanus           56         163\n",
      "139       Tanytarsus           16         119\n",
      "140    Tomosvaryella            7          37\n",
      "143      Trirhithrum           50          73\n",
      "150        Zaprionus           66          68\n",
      "48137 total number of records with 47507 records having over 500 base-pairs\n",
      "6629 records with a predefined species from which 6125 have over 500 base pairs\n",
      "\n",
      "48137 records were assigned species classification with bootstrap support ranging from 0 to 1\n",
      "\n",
      "8049 records were classiffied with a bootsrap of = 1, 9244 with >= 0.95 bootstrap support and 11337 with >= 0.7 bootstrap support\n",
      "\n",
      "4407 records, with predefined species, were classiffied with a bootsrap of = 1, 4785 with >= 0.95 bootstrap support and 5124 with >= 0.7 bootstrap support\n",
      "\n",
      "7717 records, with >= 500 seq. length, were classiffied with a bootsrap of = 1, 8883 with >= 0.95 bootstrap support and 10957 with >= 0.7 bootstrap support\n",
      "\n",
      "4122 records, with predefined species and >= 500 seq. length, were classiffied with a bootsrap of = 1, 4479 with >= 0.95 bootstrap support and 4802 with >= 0.7 bootstrap support\n",
      "\n",
      "10271 records either had predefined species, or were classified with a bootstrap value of 1.0, 11088 with bootstrap value of >=0.95 and 12842 with boostrap value of >=0.7\n",
      "\n",
      "7717 records classified into species have sequences with over 500 bases, with 3595 having no prior species classification\n",
      "422 records with a species classication confidence of 1.0 have differing assinged species from predefined species, 303 of which have over 500 base-pairs\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ~/bioinformatics/github/co1_metaanalysis/data/input/diptera\n",
    "Rscript --vanilla ../../../code/RDPresults_eval.R enafroCOI_all_clean_Diptera_taxa.tsv\n",
    "Rscript --vanilla ../../../code/RDPresults_eval.R enafroCOI_diptera_all_data_taxa.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **HEADERs**\n",
    "**Editting headers**  \n",
    "As a result of RDPclassification:  \n",
    "1. Some records will have taxons assigned to them\n",
    "2. Some will have assigned taxons that mismatch the predefined ones  \n",
    "\n",
    "To deal with this headers are generated from the RDP classifation results and attention paid to only those records with over 500 nucleotides and species assignment bootstrap of 1.0  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dipt99_headers_fasta\n",
      "diptera_clean_headers_fasta\n",
      "diptera_headers_fasta\n",
      "enafroCOI_all_clean_Diptera_95g.fasta\n",
      "enafroCOI_all_clean_Diptera_95g.fasta.reduced\n",
      "enafroCOI_all_clean_Diptera.fasta\n",
      "enafroCOI_all_clean_Diptera.fasta.reduced\n",
      "enafroCOI_all_clean_Diptera_taxa\n",
      "enafroCOI_all_clean_Diptera_taxa_clash_headers\n",
      "enafroCOI_all_clean_Diptera_taxa_clash.tsv\n",
      "enafroCOI_all_clean_Diptera_taxa_families.tsv\n",
      "enafroCOI_all_clean_Diptera_taxa_genera.tsv\n",
      "enafroCOI_all_clean_Diptera_taxa_headers\n",
      "enafroCOI_all_clean_Diptera_taxa.tsv\n",
      "enafroCOI_diptera_500to700_data-650to660.fasta\n",
      "enafroCOI_diptera_500to700_data.fasta\n",
      "enafroCOI_diptera_650to660_data.fasta\n",
      "enafroCOI_diptera_all_data.fasta\n",
      "enafroCOI_diptera_all_data_raw.fasta\n",
      "enafroCOI_diptera_all_data_taxa\n",
      "enafroCOI_diptera_all_data_taxa_clash.tsv\n",
      "enafroCOI_diptera_all_data_taxa_families.tsv\n",
      "enafroCOI_diptera_all_data_taxa_genera.tsv\n",
      "enafroCOI_diptera_all_data_taxa_headers\n",
      "enafroCOI_diptera_all_data_taxa.tsv\n",
      "enafroCOI_diptera_Over499_data.fasta\n",
      "enafroCOI_diptera_Over700_data.fasta\n",
      "enafroCOI_diptera_taxa\n",
      "enafroCOI_diptera_taxa_clash.tsv\n",
      "enafroCOI_diptera_taxa_families.tsv\n",
      "enafroCOI_diptera_taxa.tsv\n",
      "enafroCOI_diptera_Under500_data.fasta\n",
      "genera_distro_clean_95g.txt\n",
      "genera_distro.txt\n",
      "input_taxa\n",
      "input_taxa_headers\n",
      "input_taxa.tsv\n",
      "Mpala_Kinondo_Turkana_Malaise_traps_dgpd_Diptera_500to700_data-650to660.fasta\n",
      "Mpala_Kinondo_Turkana_Malaise_traps_dgpd_Diptera_650to660.fasta\n",
      "Mpala_Kinondo_Turkana_Malaise_traps_dgpd_Diptera_dgpd.fasta\n",
      "Mpala_Kinondo_Turkana_Malaise_traps_dgpd_Diptera.fasta\n",
      "Mpala_Kinondo_Turkana_Malaise_traps_dgpd_Diptera_Under500.fasta\n",
      "outgroup_coleoptera.fasta\n",
      "outgroup.fasta\n",
      "outgroup_hymenoptera_Erromyrma.fasta\n",
      "outgroup_hymenoptera.fasta\n",
      "outgroup_hymenoptera_Hypoponera.fasta\n",
      "outgroup_hymenoptera_Plagiolepis.fasta\n",
      "outgroup_hymenoptera_Tetraponera.fasta\n",
      "outgroup_lepidotera.fasta\n",
      "rooting_inputfile.fas\n",
      "rooting_inputfile.fas.reduced\n",
      "species_distro_clean_95g.txt\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ~/bioinformatics/github/co1_metaanalysis/data/input/diptera\n",
    ". ~/bioinformatics/github/co1_metaanalysis/code/classify.sh\n",
    "\n",
    "RDPcoiheaders enafroCOI_all_clean_Diptera_taxa_clash.tsv enafroCOI_all_clean_Diptera_taxa.tsv enafroCOI_diptera_all_data_taxa.tsv << EOF\n",
    "500\n",
    "1\n",
    "500\n",
    "1\n",
    "500\n",
    "1\n",
    "EOF\n",
    "\n",
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd ~/bioinformatics/github/co1_metaanalysis/data/input/diptera\n",
    ". ~/bioinformatics/github/co1_metaanalysis/code/process_all_input_files.sh\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "To delete sequences with specific words in the headers please select one of the options [1|2|3] to proceed or [4] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI222-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI222-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI174-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI174-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBMIN32883-13'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBMIN32883-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBMIN32913-13'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBMIN32913-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBDP10744-12'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBDP10744-12' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBMIN59900-17'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBMIN59900-17' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBDP10753-12'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBDP10753-12' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI223-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI223-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MAFF071-09'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MAFF071-09' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBMIN64306-17'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBMIN64306-17' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBMIN32912-13'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBMIN32912-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBCER062-14'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBCER062-14' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBMIN32837-13'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBMIN32837-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI119-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI119-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI070-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI070-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBMIN55612-17'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBMIN55612-17' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI211-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI211-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBMIN32911-13'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBMIN32911-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MAFF101-09'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MAFF101-09' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI033-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI033-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MAFF074-09'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MAFF074-09' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBMIN61347-17'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBMIN61347-17' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBMIN61354-17'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBMIN61354-17' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI293-09'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI293-09' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBDP10751-12'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBDP10751-12' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'KDIPT2474-13'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KDIPT2474-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI032-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI032-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBDP10746-12'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBDP10746-12' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'ASILO047-17'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'ASILO047-17' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI213-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI213-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'CNCDB3169-11'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'CNCDB3169-11' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'KDIPT2564-13'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KDIPT2564-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI210-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI210-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI205-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI205-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI219-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI219-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBMIN61373-17'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBMIN61373-17' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MAFF098-09'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MAFF098-09' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MAFF103-09'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MAFF103-09' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBMIN59817-17'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBMIN59817-17' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI031-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI031-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'CNCDB3171-11'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'CNCDB3171-11' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI212-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI212-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI035-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI035-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBMIN32921-13'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBMIN32921-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBMIN32882-13'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBMIN32882-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI214-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI214-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBDP10752-12'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBDP10752-12' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI029-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI029-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBMIN32852-13'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBMIN32852-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'RNOCF360-17'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'RNOCF360-17' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBMIN63209-17'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBMIN63209-17' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'CNCDB3172-11'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'CNCDB3172-11' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBMIN61374-17'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBMIN61374-17' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MAFF107-09'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MAFF107-09' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI254-09'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI254-09' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MAFF216-09'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MAFF216-09' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBDP10741-12'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBDP10741-12' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI209-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI209-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MAFF099-09'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MAFF099-09' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI220-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI220-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'KDIPT2480-13'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KDIPT2480-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI182-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI182-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MAFF075-09'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MAFF075-09' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI181-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI181-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MAFF069-09'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MAFF069-09' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MAFF100-09'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MAFF100-09' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBDP5652-09'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBDP5652-09' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI071-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI071-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'AFROM100-13'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'AFROM100-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBMIN59819-17'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBMIN59819-17' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI206-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI206-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBMIN32884-13'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBMIN32884-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'KDIPT2484-13'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KDIPT2484-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBDP10745-12'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBDP10745-12' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBMIN61350-17'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBMIN61350-17' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'ASILO059-17'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'ASILO059-17' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBDP10750-12'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBDP10750-12' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBMIN61372-17'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBMIN61372-17' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MAFF068-09'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MAFF068-09' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBDP5631-09'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBDP5631-09' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MAFF235-09'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MAFF235-09' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBDP17278-15'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBDP17278-15' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI321-09'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI321-09' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBDP17543-15'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBDP17543-15' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBDP15238-14'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBDP15238-14' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBDP10847-12'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBDP10847-12' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBDP10781-12'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBDP10781-12' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI107-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI107-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBMIN32914-13'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBMIN32914-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MAFF104-09'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MAFF104-09' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'KDIPT2481-13'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KDIPT2481-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MAFF105-09'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MAFF105-09' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBMIN32885-13'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBMIN32885-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBDP10812-12'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBDP10812-12' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBMIN32919-13'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBMIN32919-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'CNCDB3170-11'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'CNCDB3170-11' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI120-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI120-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'CNCDB3173-11'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'CNCDB3173-11' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MAFF063-09'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MAFF063-09' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI208-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI208-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI221-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI221-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBMIN61371-17'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBMIN61371-17' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MAFF102-09'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MAFF102-09' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI204-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI204-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MAFF106-09'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MAFF106-09' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBDP10743-12'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBDP10743-12' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI183-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI183-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'ASILO183-17'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'ASILO183-17' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'CNCDB1438-11'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'CNCDB1438-11' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBDP5630-09'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBDP5630-09' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBMIN32920-13'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBMIN32920-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBDP10742-12'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBDP10742-12' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBMIN32923-13'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBMIN32923-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'MVTBI235-08'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'MVTBI235-08' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern 'GBDP10817-12'\n",
      "\t...Concatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'GBDP10817-12' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "Exiting deletion of unwanted sequences...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1) Move records with word patterns specified in a file into one file\n",
      "2) Move records with word patterns specified in a file into individual word specific file\n",
      "3) Move records with specific single word into word specific file\n",
      "4) Exit\n",
      "Select option [1|2|3] to delete, or [4] to exit: Select option [1|2|3] to delete, or [4] to exit: "
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ~/bioinformatics/github/co1_metaanalysis/data/input/diptera\n",
    ". ~/bioinformatics/github/co1_metaanalysis/code/process_all_input_files.sh\n",
    "awk -F'[>|]' '/^>/{print $2}' enafroCOI_all_clean_Diptera_taxa_clash_headers > process_IDs\n",
    "\n",
    "cp enafroCOI_all_clean_Diptera.fasta infile.fasta\n",
    "delete_unwanted infile.fasta << EOF\n",
    "1\n",
    "process_IDs\n",
    "4\n",
    "EOF\n",
    "mv infile_unwanted.fasta enafroCOI_all_clean_Diptera_taxa_clash.fasta\n",
    "rm process_IDs infile.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "For the headers_[aln|fasta|fa|afa] input, provide the full path to the file, the filename included.\n",
      "\n",
      "\tStarting operation....\n",
      "\tPlease wait, this may take a while....\n",
      "\n",
      "Proceeding with enafroCOI_all_clean_Diptera_taxa_clash.fasta...\n",
      "Change for >MVTBI222-08 already in place...\n",
      "Change for >MVTBI174-08 already in place...\n",
      "Change for >GBMIN32883-13 already in place...\n",
      "Change for >GBMIN32913-13 already in place...\n",
      "Change for >GBDP10744-12 already in place...\n",
      "Change for >GBMIN59900-17 already in place...\n",
      "Change for >GBDP10753-12 already in place...\n",
      "Change for >MVTBI223-08 already in place...\n",
      "Change for >MAFF071-09 already in place...\n",
      "Change for >GBMIN64306-17 already in place...\n",
      "Change for >GBMIN32912-13 already in place...\n",
      "Change for >GBCER062-14 already in place...\n",
      "Change for >GBMIN32837-13 already in place...\n",
      "Change for >MVTBI119-08 already in place...\n",
      "Change for >MVTBI070-08 already in place...\n",
      "Change for >GBMIN55612-17 already in place...\n",
      "Change for >MVTBI211-08 already in place...\n",
      "Change for >GBMIN32911-13 already in place...\n",
      "Change for >MAFF101-09 already in place...\n",
      "Change for >MVTBI033-08 already in place...\n",
      "Change for >MAFF074-09 already in place...\n",
      "Change for >GBMIN61347-17 already in place...\n",
      "Change for >GBMIN61354-17 already in place...\n",
      "Change for >MVTBI293-09 already in place...\n",
      "Change for >GBDP10751-12 already in place...\n",
      "Change for >KDIPT2474-13 already in place...\n",
      "Change for >MVTBI032-08 already in place...\n",
      "Change for >GBDP10746-12 already in place...\n",
      "Change for >ASILO047-17 already in place...\n",
      "Change for >MVTBI213-08 already in place...\n",
      "Change for >CNCDB3169-11 already in place...\n",
      "Change for >KDIPT2564-13 already in place...\n",
      "Change for >MVTBI210-08 already in place...\n",
      "Change for >MVTBI205-08 already in place...\n",
      "Change for >MVTBI219-08 already in place...\n",
      "Change for >GBMIN61373-17 already in place...\n",
      "Change for >MAFF098-09 already in place...\n",
      "Change for >MAFF103-09 already in place...\n",
      "Change for >GBMIN59817-17 already in place...\n",
      "Change for >MVTBI031-08 already in place...\n",
      "Change for >CNCDB3171-11 already in place...\n",
      "Change for >MVTBI212-08 already in place...\n",
      "Change for >MVTBI035-08 already in place...\n",
      "Change for >GBMIN32921-13 already in place...\n",
      "Change for >GBMIN32882-13 already in place...\n",
      "Change for >MVTBI214-08 already in place...\n",
      "Change for >GBDP10752-12 already in place...\n",
      "Change for >MVTBI029-08 already in place...\n",
      "Change for >GBMIN32852-13 already in place...\n",
      "Change for >RNOCF360-17 already in place...\n",
      "Change for >GBMIN63209-17 already in place...\n",
      "Change for >CNCDB3172-11 already in place...\n",
      "Change for >GBMIN61374-17 already in place...\n",
      "Change for >MAFF107-09 already in place...\n",
      "Change for >MVTBI254-09 already in place...\n",
      "Change for >MAFF216-09 already in place...\n",
      "Change for >GBDP10741-12 already in place...\n",
      "Change for >MVTBI209-08 already in place...\n",
      "Change for >MAFF099-09 already in place...\n",
      "Change for >MVTBI220-08 already in place...\n",
      "Change for >KDIPT2480-13 already in place...\n",
      "Change for >MVTBI182-08 already in place...\n",
      "Change for >MAFF075-09 already in place...\n",
      "Change for >MVTBI181-08 already in place...\n",
      "Change for >MAFF069-09 already in place...\n",
      "Change for >MAFF100-09 already in place...\n",
      "Change for >GBDP5652-09 already in place...\n",
      "Change for >MVTBI071-08 already in place...\n",
      "Change for >AFROM100-13 already in place...\n",
      "Change for >GBMIN59819-17 already in place...\n",
      "Change for >MVTBI206-08 already in place...\n",
      "Change for >GBMIN32884-13 already in place...\n",
      "Change for >KDIPT2484-13 already in place...\n",
      "Change for >GBDP10745-12 already in place...\n",
      "Change for >GBMIN61350-17 already in place...\n",
      "Change for >ASILO059-17 already in place...\n",
      "Change for >GBDP10750-12 already in place...\n",
      "Change for >GBMIN61372-17 already in place...\n",
      "Change for >MAFF068-09 already in place...\n",
      "Change for >GBDP5631-09 already in place...\n",
      "Change for >MAFF235-09 already in place...\n",
      "Change for >GBDP17278-15 already in place...\n",
      "Change for >MVTBI321-09 already in place...\n",
      "Change for >GBDP17543-15 already in place...\n",
      "Change for >GBDP15238-14 already in place...\n",
      "Change for >GBDP10847-12 already in place...\n",
      "Change for >GBDP10781-12 already in place...\n",
      "Change for >MVTBI107-08 already in place...\n",
      "Change for >GBMIN32914-13 already in place...\n",
      "Change for >MAFF104-09 already in place...\n",
      "Change for >KDIPT2481-13 already in place...\n",
      "Change for >MAFF105-09 already in place...\n",
      "Change for >GBMIN32885-13 already in place...\n",
      "Change for >GBDP10812-12 already in place...\n",
      "Change for >GBMIN32919-13 already in place...\n",
      "Change for >CNCDB3170-11 already in place...\n",
      "Change for >MVTBI120-08 already in place...\n",
      "Change for >CNCDB3173-11 already in place...\n",
      "Change for >MAFF063-09 already in place...\n",
      "Change for >MVTBI208-08 already in place...\n",
      "Change for >MVTBI221-08 already in place...\n",
      "Change for >GBMIN61371-17 already in place...\n",
      "Change for >MAFF102-09 already in place...\n",
      "Change for >MVTBI204-08 already in place...\n",
      "Change for >MAFF106-09 already in place...\n",
      "Change for >GBDP10743-12 already in place...\n",
      "Change for >MVTBI183-08 already in place...\n",
      "Change for >ASILO183-17 already in place...\n",
      "Change for >CNCDB1438-11 already in place...\n",
      "Change for >GBDP5630-09 already in place...\n",
      "Change for >GBMIN32920-13 already in place...\n",
      "Change for >GBDP10742-12 already in place...\n",
      "Change for >GBMIN32923-13 already in place...\n",
      "Change for >MVTBI235-08 already in place...\n",
      "Change for >GBDP10817-12 already in place...\n",
      "\n",
      "DONE. 0 replacements done in enafroCOI_all_clean_Diptera_taxa_clash.fasta out of 115 records it has\n",
      "\n",
      "\tCongratulations...Operation done.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ~/bioinformatics/github/co1_metaanalysis/data/input/diptera\n",
    ". ~/bioinformatics/github/co1_metaanalysis/code/process_all_input_files.sh\n",
    "\n",
    "replacing_headers enafroCOI_all_clean_Diptera_taxa_clash.fasta << EOF\n",
    "enafroCOI_all_clean_Diptera_taxa_clash_headers\n",
    "EOF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\tRemoving gaps, '-', from infile_unwanted.fasta\n",
      "\tConcatinating sequence lines for each record in infile_unwanted.fasta...\n",
      ">KINS3436-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka_Ikuywa|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3437-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka_Ikuywa|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3438-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka_Ikuywa|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3439-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka_Ikuywa|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3440-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka_Ikuywa|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3448-13|Arthropoda|Insecta|Diptera|fam-Simuliidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka_in_the_fields|lat_0.212|lon_34.924|elev-1570.43|l-658\n",
      ">KINS3472-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-KEEP_Snake_park|lat_0.252|lon_34.92|elev-1614.17|l-658\n",
      ">KINS3473-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-KEEP_Snake_park|lat_0.252|lon_34.92|elev-1614.17|l-658\n",
      ">KINS3474-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-KEEP_Snake_park|lat_0.252|lon_34.92|elev-1614.17|l-658\n",
      ">KINS3502-13|Arthropoda|Insecta|Diptera|fam-Athericidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3503-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3504-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3505-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3506-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3507-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3617-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3618-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3619-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3620-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3621-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka|lat_0.228|lon_34.963|elev-1643.49|l-658\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ~/bioinformatics/github/co1_metaanalysis/data/input/diptera\n",
    "#removing gaps. The output is \"input_dgpd.fasta\", without gaps, \"-\".\n",
    ". ~/bioinformatics/github/co1_metaanalysis/code/process_all_input_files.sh\n",
    "remove_gaps infile_unwanted.fasta\n",
    "#Introducing a field \"l-xxx\" that has the length of the sequence in the header\n",
    "awk '/^>/{hdr=$0; next}\n",
    "    { seq=$0 } match(seq,/^.*$/) { LEN=RLENGTH }\n",
    "    { print hdr\"|l-\"LEN; print seq }' infile_unwanted_dgpd.fasta > infile_unwanted_dgpd_edited.fasta\n",
    "\n",
    "mv infile_unwanted_dgpd_edited.fasta infile_unwanted.fasta\n",
    "rm infile_unwanted_dgpd.fasta\n",
    "grep \">\" infile_unwanted.fasta > infile_headers_fasta\n",
    "head -20 infile_headers_fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd ~/bioinformatics/github/co1_metaanalysis/data/input/diptera\n",
    "#substituting headers from infile_headers_fasta into all *diptera*.fasta files.\n",
    ". ~/bioinformatics/github/co1_metaanalysis/code/process_all_input_files.sh\n",
    "replacing_headers diptera_clean_headers_fasta *diptera*.fasta << EOF\n",
    "infile_headers_fasta\n",
    "EOF\n",
    "#cleaning up residual files\n",
    "mv infile_headers_fasta dipt99_headers_fasta\n",
    "rm -rf infile.fasta infile_unwanted.fasta outputfile process_IDs sedeW4xeJ"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
